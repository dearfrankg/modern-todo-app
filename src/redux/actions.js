import * as types from 'redux/types';

export const addTodo = text => ({ type: types.ADD_TODO, text });
export const deleteTodo = id => ({ type: types.DELETE_TODO, id });
export const editTodo = (id, text) => ({ type: types.EDIT_TODO, id, text });
export const toggleTodo = todo => ({ type: types.TOGGLE_TODO, todo });
export const completeAll = () => ({ type: types.COMPLETE_ALL });
export const clearCompleted = () => ({ type: types.CLEAR_COMPLETED });
