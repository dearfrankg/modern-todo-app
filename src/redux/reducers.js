import * as types from 'redux/types';

export const initialState = {
  todos: [],
};

const todos = (state = initialState.todos, action) => {
  if (!action) return false;

  switch (action.type) {
    case types.ADD_TODO: {
      const maxId =
        state.length === 0
          ? 0
          : state.reduce((prev, current) => (prev.id > current.id ? prev : current)).id;

      return [
        {
          id: maxId + 1,
          text: action.text,
          completed: false,
        },
        ...state,
      ];
    }

    case types.DELETE_TODO: {
      return state.filter(todo => todo.id !== action.id);
    }

    case types.EDIT_TODO: {
      return state.map((todo) => {
        if (todo.id === action.id) return { ...todo, text: action.text };

        return todo;
      });
    }

    case types.TOGGLE_TODO: {
      return state.map((todo) => {
        if (todo.id === action.todo.id) return { ...todo, completed: !action.todo.completed };

        return todo;
      });
    }

    case types.COMPLETE_ALL: {
      const areAllMarked = state.every(todo => todo.completed);
      return state.map(todo => ({
        ...todo,
        completed: !areAllMarked,
      }));
    }

    case types.CLEAR_COMPLETED: {
      return state.filter(todo => todo.completed === false);
    }

    default:
      return state;
  }
};

export default todos;
