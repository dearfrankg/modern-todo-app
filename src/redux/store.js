import { createStore, combineReducers } from 'redux';
import { routerReducer } from 'react-router-redux';

import rootReducer from 'redux/reducers';
import { loadState, saveState } from 'redux/localStorage';

const persistedState = loadState();

const store = createStore(
  combineReducers({
    todos: rootReducer,
    router: routerReducer,
  }),
  persistedState,
);

store.subscribe(() => {
  saveState({
    todos: store.getState().todos,
  });
});

export default store;
