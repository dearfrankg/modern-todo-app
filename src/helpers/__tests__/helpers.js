import * as filters from 'redux/types';

import { getListStatus, getAllCompleted, getItemsLeftLabel, getFilterFromPath } from '../index';

const todo = {
  id: 1,
  completed: false,
  text: 'Todo',
};

describe('Helpers:', () => {
  describe('getListStatus():', () => {
    it('should return false when errors|undefined', () => {
      expect(getListStatus()).toBe(false);
    });

    it('should return the todo list with status correctly with 1 completed todo', () => {
      expect(getListStatus([{ completed: false }])).toHaveProperty('notCompletedItems', 1);
    });

    it('should return the todo list with status correctly with 1 not completed todo', () => {
      expect(getListStatus([{ completed: true }])).toHaveProperty('completedItems', 1);
    });

    it('should return the todo list with status correctly with severeal todos with different status', () => {
      expect(
        getListStatus([{ completed: false }, { completed: true }, { completed: false }]),
      ).toMatchObject({ completedItems: 1, notCompletedItems: 2 });
    });
  });

  describe('getAllCompleted():', () => {
    it('should return false when errors|undefined', () => {
      expect(getAllCompleted()).toBe(false);
    });

    it('should return all completed correctly when no items completed', () => {
      expect(getAllCompleted([todo, todo])).toEqual(false);
    });

    it('should return all completed correctly when has some completed', () => {
      expect(getAllCompleted([todo, { ...todo, completed: true }])).toEqual(false);
    });

    it('should return all completed correctly when all completed', () => {
      expect(getAllCompleted([{ ...todo, completed: true }, { ...todo, completed: true }])).toEqual(
        true,
      );
    });
  });

  describe('getItemsLeftLabel():', () => {
    it('should return false when errors|undefined', () => {
      expect(getItemsLeftLabel()).toBe(false);
    });

    it('should return correctly when is it has more than 1 todo', () => {
      expect(
        getItemsLeftLabel([{ completed: false }, { completed: true }, { completed: false }]),
      ).toMatchSnapshot();
    });

    it('should return correctly when is it has just a single todo', () => {
      expect(getItemsLeftLabel([{ completed: false }, { completed: true }])).toMatchSnapshot();
    });

    it('should return correctly when is it has no todos', () => {
      expect(getItemsLeftLabel([{ completed: true }, { completed: true }])).toMatchSnapshot();
    });
  });

  describe('getFilterFromPath():', () => {
    it('should return false when errors|undefined', () => {
      expect(getFilterFromPath()).toBe(false);
    });

    it('should return correctly when it has hash but not match any pattenrs', () => {
      expect(getFilterFromPath('test')).toBe(filters.SHOW_ALL);
    });

    it('should return correctly when show_all', () => {
      expect(getFilterFromPath('#/')).toBe(filters.SHOW_ALL);
    });

    it('should return correctly when show_active', () => {
      expect(getFilterFromPath('#/active')).toBe(filters.SHOW_ACTIVE);
    });

    it('should return correctly when show_completed', () => {
      expect(getFilterFromPath('#/completed')).toBe(filters.SHOW_COMPLETED);
    });
  });
});
