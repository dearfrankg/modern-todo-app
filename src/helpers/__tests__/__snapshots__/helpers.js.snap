// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Helpers: getItemsLeftLabel(): should return correctly when is it has just a single todo 1`] = `
<span>
  <strong>
    1
  </strong>
   item left
</span>
`;

exports[`Helpers: getItemsLeftLabel(): should return correctly when is it has more than 1 todo 1`] = `
<span>
  <strong>
    2
  </strong>
   items left
</span>
`;

exports[`Helpers: getItemsLeftLabel(): should return correctly when is it has no todos 1`] = `
<span>
  0 items left
</span>
`;
